src/contracts/reader/GPv2TradeSimulator.sol

```diff
@@ -0,0 +1,135 @@
+// SPDX-License-Identifier: LGPL-3.0-or-later
+pragma solidity ^0.7.6;
+pragma abicoder v2;
+
+import "../GPv2Settlement.sol";
+
+/// @title Gnosis Protocol v2 Trade Simulator
+/// @author Gnosis Developers
+contract GPv2TradeSimulator {
+    using GPv2Transfer for IVault;
+    using SafeCast for int256;
+    using SafeCast for uint256;
+
+    /// @dev A trade to simulate.
+    struct Trade {
+        address owner;
+        address receiver;
+        IERC20 sellToken;
+        IERC20 buyToken;
+        uint256 sellAmount;
+        uint256 buyAmount;
+        bytes32 sellTokenBalance;
+        bytes32 buyTokenBalance;
+    }
+
+    /// @dev Initial token state used internally for computing balance changes.
+    struct TokenState {
+        IERC20 token;
+        uint256 initialBalance;
+    }
+
+    /// @dev Simulation context.
+    struct Context {
+        GPv2Transfer.Data[] inTransfers;
+        GPv2Transfer.Data[] outTransfers;
+        TokenState sellTokenState;
+        TokenState buyTokenState;
+    }
+
+    /// @dev Sentinal value to indicate that the buy amount for the trade's out
+    /// transfer should use all received buy tokens from the interactions
+    /// specified for the simulation.
+    uint256 private constant USE_ALL_RECEIVED_BUY_TOKENS = 0;
+
+    /// @dev Simulates a user trade.
+    ///
+    /// This method can be used to determine whether or not a token is supported
+    /// as well as getting a rough estimate on how much gas is required to
+    /// execute the trade given a set of interactions.
+    ///
+    /// One notable difference, is the simulation doesn't actually require an
+    /// order. This makes using this method inaccurate for predicting gas usage
+    /// for specific trades. However, seeing as computing trade executions is
+    /// completely independant of external contracts (i.e. it does not depend on
+    /// the tokens being traded or the interactions being executed) the gas
+    /// consumption should be off by a predicatable amount.
```

Comment:
This is unfortunately not true for EIP1271 signatures, for which we might eventually need a different gas estimation method...

Reply:
I imagine we would have to verify the EIP-1271 signature when adding he order, which would give us a gas estimate right?

Reply:
I see, I imagine the `quote` API will need to change to reflect this then.